enhancements/mcd-reboot-dictionary.md

```diff
@@ -0,0 +1,215 @@
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+- Should we allow the whitelist to apply recursively to a directory, or make
+  use of pattern matching?
+
+- Is there any way to mitigate the version skew problem?
+
+- Should the parts of the configuration that do and don't require reboot be
+  part of user facing documentation?
+
+
+## Summary
+
+A node's configuration is managed by the MCO.  When the configuation is changed
+by the system or an admin, the MCO and MCD write a new version of the file to
+disk and reboot the system to ensure it is applied.
+
+This proposal is to create a dictionary that defines a whitelist of file paths
+for which alternative application strategies should be used.  
+
+
+## Motivation
+
+Always rebooting a node is reasonable for cloud based systems, since they can
+be rebooted very quickly.  Additionally, it is the only way for kernel
+parameter changes to take effect.
+
+However on baremetal systems, rebooting a node takes significantly longer (in
+the order of minutes) and the machine configuration includes files that need
+either no action to take effect, or can be more quickly applied by restarting a
+systemd service.
+
+### Goals
+
+- Creation of a minimally invasive solution that can be delivered in the 
+  short-term, allowing time for a more comprehensive approach to be developed
+- A system defined whitelist of file paths that do not trigger a reboot
+- Changes to systemd service configurations are applied by restarting the service
+- Specific files can be updated with no action (eg. ssh keys)
+- Any file not covered by the whitelist triggers a reboot
+
+### Non-Goals
+
+- Allowing the admin to specify additional paths or strategies
+- Allowing different behaviour based on which option or part of the file has
+  been updated
+
+## Proposal
+
+The proposal calls for a whitelist that MCD uses to decide which action to take
+after updating a configuration file for the MCO.  To prevent the admin from
+tampering with the whitelist, it will be included as part of the MCD image as a
+static file in yaml format.
+
+The information could be baked into the the MCD itself, but assuming we treat it
+as data, the file will be a list of entries containing:
+
+- filename
+- path
+- action to perform
+- action specific data (eg. the service name, timeout)
+
+After the MCD writes out a configuation change, it will consult the whitelist
+before deciding if a reboot is required. If the filename is present, has a
+valid action and any associated data, the action will be performed.  Otherwise,
+a reboot will be performed as in prior 4.x versions.
+
+If the action results in an error, or takes too long, a reboot will be
+performed as in prior 4.x versions.
```

Comment:
This section is not clear..

- Which errors are we talking about, file replace error, unit restart errors ??
- Also timeouts are tricky.. is MCD supposed to max the timeouts from multiple whiteliests? When should MCD start waiting? which specific timeouts are being talked about here?

I think if we can default these in detail wrt files, unit files, file that affect units, kernel args etc.. that would be better.

Reply:
ack

Reply:
I've dropped the timeout from the file list.
We may eventually want to put a cap on how long drains can take, but exposing it as a per-entry tuneable requires too much state.